rfcs/20210510-sum-with-zero-point/README.md

```diff
@@ -0,0 +1,158 @@
+            else
+                if (zero_point == 0)
+                    error::wrap_c_api(dnnl_post_ops_append_sum_v2(
+                                              get(), scale,
+                                              memory::convert_to_c(data_type)),
+                            "could not append a sum post-op");
+                else
+                    error::wrap_c_api(dnnl_post_ops_append_sum_v3(
+                                              get(), scale,
+                                              memory::convert_to_c(data_type)),
+                                              zero_point,
+                            "could not append a sum post-op");
+        }
+        ```
+    - Pros:
+        - It is backwards compatible.
+        - Allows to use zero point without scale (scale == 1.0).
+    - Cons:
+        - It is not the simplest solution
+    - Example:
+        ```cpp
+        // Zero point is the last parameter to keep existing code working.
+        attr.append_sum(scale, dt);
+        attr.append_sum(scale, dt, zp);
+        ```
+
+2. Extend existing `append_sum(scale, dt)` and `dnnl_post_ops_append_sum_v2` to
+support zero point.
+    - API:
+        ```cpp
+        // include/oneapi/dnnl/dnnl.h
+        dnnl_status_t DNNL_API dnnl_post_ops_append_sum_v2(
+                dnnl_post_ops_t post_ops, float scale, int32_t zero_point,
+                dnnl_data_type_t data_type);
+
+        // include/oneapi/dnnl/dnnl.hpp
+            void append_sum(float scale = 1.f, int32_t zero_point = 0,
+                    memory::data_type data_type = memory::data_type::undef) {
+                if (data_type == memory::data_type::undef)
+                    error::wrap_c_api(dnnl_post_ops_append_sum(get(), scale),
+                            "could not append a sum post-op");
+                else
+                    error::wrap_c_api(dnnl_post_ops_append_sum_v2(
+                                              get(), scale, zero_point,
+                                              memory::convert_to_c(data_type)),
+                            "could not append a sum post-op");
+            }
+        ```
+    - Pros:
+        - Very simple solution.
+        - Universal solution.
+    - Cons:
+        - It is not backwards compatible for C API.
+    - Example:
+        ```cpp
+        // Zero point is the last parameter to keep existing code working.
+        attr.append_sum(scale, dt);
+        ```
+
+3. Add possibility to use `dnnl_primitive_attr_set_zero_points` with sum.
+    - Pros:
+        - It will allow to use mask and array of zero-point values (instead of
+          one scalar value).
+    - Cons:
+        - More complicated solution.
+        - Probably requires backward incompatibility or adding a new function.
+        - Probably not needed in real topologies.
+    - Example:
+        ```cpp
+        attr.set_zero_points(DNNL_ARG_SRC_1, mask, zero_points);
```

Comment:
```suggestion
        attr.set_zero_points(DNNL_ARG_ATTR_MULTIPLE_POST_OP(idx), mask, zero_points);
```

This is the proper mecanism to pass attributest that are specific to post-ops.

Reply:
@mgouicem What do you think about the following version:
```cpp
attr.set_zero_points(DNNL_ARG_ATTR_MULTIPLE_POST_OP(idx) | DNNL_ARG_SRC_1, mask, zero_points);
```

Reply:
It depends on the exact semantic of the formula I guess :-).
If the formula does not change (so scale/zero point always apply to dst), then no need to specify the argument kind.
If the formula changes to allow applying scale/zero_point to `op()`, then yes, it makes sense to specify the argument kind.

Though I would refrain from changing the formula, since:
a. it does not seem useful 
b. even if a user wants to do that, it could be achieved with eltwise linear postop.

Reply:
Here, the argument kind specified for zero_point is relative to the primitive arguments.
So if we use `DNNL_ARG_SRC1`, the zero point would apply to `pd.src_desc(1)` of a given primitive descriptor.
In particular, this will be ignored for almost all primitives other than concat, sum and binary.

In the context of the sum postop (or postops in general), we want to pass a zero_point to will be applied during the sum postop execution. To achieve that, we need to specify that the zero point will apply in the given postop (achieved with `DNNL_ARG_ATTR_MULTIPLE_POST_OP(idx)`). If that postop can take different zero_point for its arguments, then we need to disambiguate by specifying which argument as well (e.g. `DNNL_ARG_ATTR_MULTIPLE_POST_OP(idx) | DNNL_ARG_SRC1`).

Reply:
Understand, changed.